rfd/0096-securing-docker-images.md

```diff
@@ -0,0 +1,378 @@
+    build_ok?{Does all go well<br/>with the rest of<br/>the build?}
+    sbom[Generate SBOM with syft]
+    attach_sbom[Attach SBOM to<br/>Candidate Image with cosign]
+    
+    pull_base --> signature_valid?
+    signature_valid? -- yes --> base    
+    signature_valid? -- no --> fail_build
+    teleport_deb --> build
+    third_party_debs --> build
+    df --> build
+    base --> build    
+    build --> candidate_image
+
+    candidate_image --> smoke_test   
+    smoke_test --> smoke_test_pass?
+    smoke_test_pass? -- yes --> push_internal
+    smoke_test_pass? -- no --> fail_build
+    push_internal --> trivy_scan
+    push_internal --> sbom
+    trivy_scan --> upload_results 
+    push_internal --> sign_image
+    sbom --> attach_sbom
+    upload_results --> code_scanning_alerts[(GitHub Code<br/>Scanning alerts)]
+    code_scanning_alerts --> panther[(Panther SIEM)]
+    
+    sign_image --> build_ok?
+    upload_results --> build_ok?
+    attach_sbom --> build_ok?
+    
+    promote_build --> promote_image
+    build_ok? -- yes --> promote_build
+    build_ok? -- no --> fail_build
+```    
+
+### Debug Images
+
+Troubleshooting a distroless image is hard, as there are no tools baked into
+the image to aid in debugging a deployment.
+
+In addtion to the main distroless image sets, the `distroless` team also supplies 
+a `debug`-tagged image that includes `busybox` (that is, a basic shell and some
+utilities). In order to provide tooling for troubleshooting a Teleport installation,
+we will create a parallel `teleport-debug` image, based on a distroless `debug` image.
+
+While we should take as much care as possible when constructing and 
+monitoring this image, use of the debug image should probably be
+considered "at your own risk".
+
+### Compatibility Guarantees
+
+We have clients relying on the existing behaviour (and contents) of our images. We
+should treat releasing these distroless images as a compatibility break, and make 
+our customers aware of our intentions well in advance so that they can prepare.
+
+## 2. Maintaininece 
+
+### Collecting data 
+
+We are already using the scanning tools built-in to AWS ECR, which periodically scans our 
+images and results of the scan into our Panther SIEM instance.
+
+To increase the size and quality of the vulnetrability database we scan with, we should 
+also use `trivy` for ongoing scans of our published (i.e. public) images, in addition
+to the build-time scans described above. 
+
+Trivy's in a GHA integration should make it straightforward to run 
+workflow to repeatedly scan our published images on a regular schedule.
+
+The output of the scan will be injected into the GitHub code scanning system. From there 
+the alerts can be picked up buy our Panther SIEM, which already integrates into our 
+GitHub account. 
+
+### Reacting to alerts
+
+Once the alert data is aggregated into Panther, we can configure events to lodge 
```

Comment:
@reedloden @adaadb6 @wadells, Thoughts on if the security team should triage these alerts first, or should the dev team?

I prefer these alerts shift left to the dev team first, but Panther may not be the easiest tool to do that.  Thoughts?

Reply:
Just to follow up on this, it is doable with Panther. We don't have a dedicated SOAR but we can send alerts base on tags or certain attributes to a specific channel.

Reply:
Hmmm... lets consider our different container use cases:

* **Vuln found in a not-yet-published image**:  This should be shifted left, and ideally the dev team.  However we still need to consider the SLA and compensating factors -- so we probably can't have a hardline "nothing ships with a known vuln policy -- there needs to be an override mechanism to mark some automation flagged vulns as acceptable.
* **New vuln found in a previously published image**:
  * If the image is one of our fully versioned images (e.g. `teleport:12.0.2`), we currently do not change these artifacts after they're published, so there is nothing to be done. Public ECR doesn't allow tag immutability, but our private ECR where these are staged has tag immutablity turned on -- preventing these from ever being updated.
  * If the image is one of our floating tags (e.g. `teleport:12`), we want to know and make sure the vuln is fixed within the SLA.  I'd imagine we _want_ to treat these like any other vulnerability tracked in the security vulnerabilities project.  Having it go through Panther and to us first seems fine to me.

* **If the image is a `debug` image**: We can probably skip scanning and alerting on these.  They're expected to have a lot more packages and are intentionally classified as less secure.

In short, a bit of column A and a bit of column B